common/flatpak-utils-http.c

```diff
@@ -22,22 +22,307 @@
-  GString               *content;
   char                   buffer[16 * 1024];
   FlatpakLoadUriProgress progress;
   GCancellable          *cancellable;
   gpointer               user_data;
   guint64                last_progress_time;
+  CacheHttpData         *cache_data;
   char                  *etag;
 } LoadUriData;
 
+#define CACHE_HTTP_XATTR "user.flatpak.http"
+#define CACHE_HTTP_SUFFIX ".flatpak.http"
+#define CACHE_HTTP_TYPE "(sstt)"
+
+static void
+clear_cache_http_data (CacheHttpData *data,
+                       gboolean       clear_uri)
+{
+  if (clear_uri)
+    g_clear_pointer (&data->uri, g_free);
+  g_clear_pointer (&data->etag, g_free);
+  data->last_modified = 0;
+  data->expires = 0;
+}
+
+static void
+free_cache_http_data (CacheHttpData *data)
+{
+  clear_cache_http_data (data, TRUE);
+  g_free (data);
+}
+
+G_DEFINE_AUTOPTR_CLEANUP_FUNC (CacheHttpData, free_cache_http_data)
+
+static GBytes *
+serialize_cache_http_data (CacheHttpData * data)
+{
+  g_autoptr(GVariant) cache_variant = NULL;
+
+  cache_variant = g_variant_ref_sink (g_variant_new (CACHE_HTTP_TYPE,
+                                                     data->uri,
+                                                     data->etag ? data->etag : "",
+                                                     data->last_modified,
+                                                     data->expires));
+  if (G_BYTE_ORDER != G_BIG_ENDIAN)
+    {
+      g_autoptr(GVariant) tmp_variant = cache_variant;
+      cache_variant = g_variant_byteswap (tmp_variant);
+    }
+
+  return g_variant_get_data_as_bytes (cache_variant);
+}
+
+static void
+deserialize_cache_http_data (CacheHttpData *data,
+                             GBytes        *bytes)
+{
+  g_autoptr(GVariant) cache_variant = NULL;
+
+  cache_variant = g_variant_ref_sink (g_variant_new_from_bytes (G_VARIANT_TYPE (CACHE_HTTP_TYPE),
+                                                                bytes,
+                                                                FALSE));
+  if (G_BYTE_ORDER != G_BIG_ENDIAN)
+    {
+      g_autoptr(GVariant) tmp_variant = cache_variant;
+      cache_variant = g_variant_byteswap (tmp_variant);
+    }
+
+  g_variant_get (cache_variant,
+                 CACHE_HTTP_TYPE,
+                 &data->uri,
+                 &data->etag,
+                 &data->last_modified,
+                 &data->expires);
+}
+
+static CacheHttpData *
+load_cache_http_data (int           dfd,
+                      char         *name,
+                      gboolean     *no_xattr,
+                      GCancellable *cancellable,
+                      GError      **error)
+{
+  g_autoptr(CacheHttpData) data = NULL;
+
+  g_autoptr(GBytes) cache_bytes = glnx_lgetxattrat (dfd, name,
+                                                    CACHE_HTTP_XATTR,
+                                                    error);
+  if (cache_bytes == NULL)
+    {
+      if (errno == ENOTSUP)
```

Comment:
I checked glnx, and it does indeed ensure errno is kept correct after setting the GError. Still, shouldn't we just a g_error_matches() here?